# Кузнецов Даниил 3821Б1ПР1

## 38. Выплаты по кредиту. Решение трансцендентного уравнения

Предположим, что кто-то желает взять кредит на сумму . Он выбирает среди следующих возможностей:

1) взять кредит на  лет с ежемесячными выплатами в ;

2) взять кредит на  лет с ежемесячными выплатами в ;

3) взять кредит на  лет с ежемесячными выплатами в .

Какой кредит выгоднее брать?

Можно просто вычислить переплату, а можно еще учесть процентные ставки.

Сами процентные ставки здесь не указаны, но можно их найти из основного финансового уравнения:
${\rm PV}\, (1+i)^n + {\rm PMT}\, (1+i)^{n-1} + {\rm PMT}\, (1+i)^{n-2} + \dots + {\rm PMT} + {\rm FV} = 0, \tag{1} $

связывающего следующие величины:

 PV - величина кредита (present value);
 FV - баланс после  платежей (future value);
 PMT - величина периодического платежа со знаком минус (payment);
 n - количество платежей;
 i - процентная ставка (interest rate).
Например, для случая 1):
PV = 100000;
FV = 0;
PMT = -1500;
n = 10 * 12 = 120

Используя формулу для суммы элементов геометрической прогрессии, уравнение (1) запишем в следующем виде:

$ {\rm PV}\, (1+i)^n + {\rm PMT}\, \frac{(1+i)^n - 1}{i} + {\rm FV} = 0, \tag{2} $

Относительно  это уравнение в общем случае аналитические решено быть не может, поэтому надо использовать численные методы.

1) Написать функцию, решающее уравнение (2). Ваша функция должна вызывать scipy.optimize.brentq. Можете воспользоваться другим методом. Обоснуйте свой выбор.

2) Рассчитать годовые процентные ставки для примеров 1)-3). Годовая процентная ставка равна i_year = m * i, где m - количество платежей в год (в нашем случае m = 12).

3) Сравнить процентные ставки. Какой кредитный план выбрали бы вы?

Импортируем библиотеки:

In [77]:
import numpy as np
import scipy as sp
from scipy.optimize import brentq

Добавляем величину кредита pv и баланс после n платежей(future value) и количество платежей в год(m)

In [78]:
pv = 100000
fv = 0
m = 12

Добавляем переменные количества месяцев(выплат) и величины периодических платежей со знаком минус

In [79]:
n_fst = 10*m
n_snd = 15*m
n_trd = 20*m
pmt1 = -1500
pmt2 = -1250
pmt3 = -1160


Функция решения этого уравнения через метод Брента
$$ {\rm PV}\, (1+i)^n + {\rm PMT}\, \frac{(1+i)^n - 1}{i} + {\rm FV} = 0, \tag{2} $$


In [99]:
def month_percent (pv, fv, n, pmt):
    def finance(i):
        return pv*(1+i)**n + pmt*((((1+i)**n)-1)/i) + fv
    return brentq(finance, 0.01, 1.0)


Далее находим месячные процентные ставки для каждого вариант кредита

In [100]:
i1 = month_percent(pv, fv, n_fst, pmt1)
print(i1)

0.010930603939868333


In [101]:
i2 = month_percent(pv, fv, n_snd, pmt2)
print(i2)

0.01063980717924384


In [102]:
i3 = month_percent(pv, fv, n_trd, pmt3)
print(i3)

0.010697698792848072


Рассчитываем годовые процентные ставки:

In [103]:
i_year1 = i1 * m
i_year2 = i2 * m 
i_year3 = i3 * m 

In [104]:
print("Годовые ставки: ")
print("Первый вариант: ", round(i_year1 * 100, 2), "%")
print("Второй вариант: ", round(i_year2 * 100, 2), "%")
print("Третий вариант: ", round(i_year3 * 100, 2), "%")

Годовые ставки: 
Первый вариант:  13.12 %
Второй вариант:  12.77 %
Третий вариант:  12.84 %


Переплаты составят:

In [105]:
#Первый вариант
over1 = -pmt1 * n_fst - pv
print("Первый вариант: ", over1)
#Второй вариант
over2 = -pmt2 * n_snd - pv
print("Второй вариант: ", over2)
#Третий вариант
over3 = -pmt3 * n_trd - pv
print("Третий вариант: ", over3)

Первый вариант:  80000
Второй вариант:  125000
Третий вариант:  178400


# Вывод

Исходя из полученных годовых ставок можно сделать вывод, что наиболее выгодным кредитным планом будет второй, который обладает наименьшей процентной ставкой в 12.77%.
Использовался метод Брента так как он обладает надежными и быстрыми свойствами сходимости.
